File: app/src/main/java/com/nimbl3/having/exchange/ui/activity/ActivityChat.java
package com.nimbl3.having.exchange.ui.activity;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.text.format.DateFormat;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;

import com.firebase.ui.database.FirebaseListAdapter;
import com.google.firebase.database.FirebaseDatabase;
import com.nimbl3.having.exchange.R;
import com.nimbl3.having.exchange.ui.model.ChatMessage;

/**
 * Created by thuypham on 3/15/2018 AD.
 */

public class ActivityChat extends ActivityBase {

    private EditText mChatEdt;
    private Button mBtnSubmit;
    private FirebaseListAdapter<ChatMessage> adapter;
    private Toolbar mToolbar;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);

        mChatEdt = (EditText) findViewById(R.id.edt_chat);
        mBtnSubmit = (Button) findViewById(R.id.btn_submit);
        mToolbar = findViewById(R.id.toolbar);
        setSupportActionBar(mToolbar);
        mBtnSubmit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mChatEdt.getText().toString().trim().length() > 0) {
                    // Read the input field and push a new instance
                    // of ChatMessage to the Firebase database
                    String user = getUser();
                    FirebaseDatabase.getInstance()
                        .getReference("chats")
                        .push()
                        .setValue(new ChatMessage(mChatEdt.getText().toString(), "user_name" + user));

                    // Clear the input
                    mChatEdt.setText("");
                }
            }
        });

        ListView listOfMessages = (ListView) findViewById(R.id.list_chat);

        adapter = new FirebaseListAdapter<ChatMessage>(this, ChatMessage.class,
            R.layout.message, FirebaseDatabase.getInstance().getReference("chats")) {
            @Override
            protected void populateView(View v, ChatMessage model, int position) {
                // Get references to the views of message.xml
                TextView messageText = v.findViewById(R.id.message_text);
                TextView messageUser = v.findViewById(R.id.message_user);
                TextView messageTime = v.findViewById(R.id.message_time);

                // Set their text
                messageText.setText(model.getMessageText());
                messageUser.setText(model.getMessageUser());

                // Format the date before showing it
                messageTime.setText(DateFormat.format("dd-MM-yyyy (HH:mm:ss)",
                    model.getMessageTime()));
            }
        };

        listOfMessages.setAdapter(adapter);
    }

    private String getUser() {
        SharedPreferences sharedPreferences = this.getSharedPreferences(getString(R.string.app_name), Context.MODE_PRIVATE);
        return sharedPreferences.getString("id", "0");
    }
}

File: app/src/main/java/com/nimbl3/having/exchange/ui/activity/ActivityChat.kt
package com.nimbl3.having.exchange.ui.activity

import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.support.v7.widget.Toolbar
import android.text.format.DateFormat
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.ListView
import android.widget.TextView
import com.firebase.ui.database.FirebaseListAdapter
import com.google.firebase.database.FirebaseDatabase
import com.jakewharton.rxbinding2.view.RxView
import com.nimbl3.having.exchange.R
import com.nimbl3.having.exchange.ui.intents.ChatIntents
import com.nimbl3.having.exchange.ui.model.ChatMessage
import com.nimbl3.having.exchange.ui.viewmodel.ChatViewModel
import com.nimbl3.having.exchange.ui.viewstate.chat.ChatClearInputTextViewState
import com.nimbl3.having.exchange.ui.viewstate.chat.ChatEmptyViewState
import com.nimbl3.having.exchange.ui.viewstate.chat.ChatViewState
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable

class ActivityChat : ActivityBase() {

    lateinit var edtChat: EditText
    lateinit var btnSubmit: Button
    private var adapter: FirebaseListAdapter<ChatMessage>? = null
    private var mToolbar: Toolbar? = null
    lateinit var mViewModel: ChatViewModel
    lateinit var listOfMessages: ListView
    lateinit var emptyView: View
    lateinit var mDisposables: CompositeDisposable

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)

        edtChat = findViewById<View>(R.id.edt_chat) as EditText
        btnSubmit = findViewById<View>(R.id.btn_submit) as Button
        mToolbar = findViewById(R.id.toolbar)
        emptyView = findViewById(R.id.empty_view)

        setSupportActionBar(mToolbar)

        mViewModel = ViewModelProviders.of(this).get<ChatViewModel>(ChatViewModel::class.java);
        mDisposables = CompositeDisposable()
        bindToViewModel()

        listOfMessages = findViewById<View>(R.id.list_chat) as ListView

        adapter = object : FirebaseListAdapter<ChatMessage>(this, ChatMessage::class.java,
                R.layout.message, FirebaseDatabase.getInstance().getReference("chats")) {
            override fun populateView(v: View, model: ChatMessage, position: Int) {
                // Get references to the views of message.xml
                val messageText = v.findViewById<TextView>(R.id.message_text)
                val messageUser = v.findViewById<TextView>(R.id.message_user)
                val messageTime = v.findViewById<TextView>(R.id.message_time)

                // Set their text
                messageText.text = model.messageText
                messageUser.text = model.messageUser

                // Format the date before showing it
                messageTime.text = DateFormat.format("dd-MM-yyyy (HH:mm:ss)",
                        model.messageTime)
            }
        }

        listOfMessages.adapter = adapter
    }

    private fun bindToViewModel() {
        // Subscribe to the ViewModel and call render for every emitted state
        mDisposables.add(mViewModel.states().subscribe({ this.render(it) }))
        // Pass the UI's intents to the ViewModel
        mViewModel.processIntents(intents())
    }

    private fun render(it: ChatViewState?) {
        when (it) {
            is ChatClearInputTextViewState -> {
                renderClearInputTextState()
            }
            is ChatEmptyViewState -> {
                renderEmptyViewState()
            }
        }
    }

    private fun renderEmptyViewState() {
        emptyView.visibility = View.VISIBLE
    }

    private fun renderClearInputTextState() {
        edtChat.setText("")
    }

    private fun initialIntent(): Observable<ChatIntents.InitialIntents> {
        return Observable.just(ChatIntents.InitialIntents())
    }

    private fun submitChatIntent(): Observable<ChatIntents.SubmitChatIntent> {
        return RxView
                .clicks(btnSubmit)
                .map { _ -> ChatIntents.SubmitChatIntent(edtChat.text.toString().trim(), getUser()) }
    }

    private fun intents(): Observable<ChatIntents> {
        return Observable.merge(initialIntent(), submitChatIntent())
    }

    private fun getUser(): String {
        val sharedPreferences = this.getSharedPreferences(getString(R.string.app_name), Context.MODE_PRIVATE)
        return sharedPreferences.getString("id", "0")
    }
}
